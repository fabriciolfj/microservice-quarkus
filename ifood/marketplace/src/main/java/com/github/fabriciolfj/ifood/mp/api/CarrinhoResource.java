package com.github.fabriciolfj.ifood.mp.api;

import com.github.fabriciolfj.ifood.mp.api.dto.PedidoRealizadoDTO;
import com.github.fabriciolfj.ifood.mp.api.dto.PratoDTO;
import com.github.fabriciolfj.ifood.mp.api.dto.PratoPedidoDTO;
import com.github.fabriciolfj.ifood.mp.api.dto.RestauranteDTO;
import com.github.fabriciolfj.ifood.mp.domain.model.Prato;
import com.github.fabriciolfj.ifood.mp.domain.model.PratoCarrinho;
import io.smallrye.mutiny.Uni;
import io.vertx.mutiny.pgclient.PgPool;
import org.eclipse.microprofile.reactive.messaging.Channel;
import org.eclipse.microprofile.reactive.messaging.Emitter;

import javax.inject.Inject;
import javax.ws.rs.Consumes;
import javax.ws.rs.GET;
import javax.ws.rs.POST;
import javax.ws.rs.Path;
import javax.ws.rs.PathParam;
import javax.ws.rs.Produces;
import javax.ws.rs.core.MediaType;
import java.math.BigDecimal;
import java.util.List;
import java.util.stream.Collectors;

@Path("carrinho")
@Produces(MediaType.APPLICATION_JSON)
@Consumes(MediaType.APPLICATION_JSON)
public class CarrinhoResource {

    private static final String CLIENTE = "a";

    @Inject
    PgPool client;

    @Inject
    @Channel("pedidos")
    Emitter<PedidoRealizadoDTO> emitterPedido;

    @GET
    public Uni<List<PratoCarrinho>> buscarCarrinho() {
        return PratoCarrinho.findCarrinho(client, CLIENTE);
    }

    @POST
    @Path("/prato/{idPrato}")
    public Uni<Long> adicionarPrato(@PathParam("idPrato") Long idPrato) {
        //poderia retornar o pedido atual
        PratoCarrinho pc = new PratoCarrinho();
        pc.cliente = CLIENTE;
        pc.prato = idPrato;
        return PratoCarrinho.save(client, CLIENTE, idPrato);

    }

    @POST
    @Path("/realizar-pedido")
    public Uni<Boolean> finalizar() {
        PedidoRealizadoDTO pedido = new PedidoRealizadoDTO();
        String cliente = CLIENTE;
        pedido.cliente = cliente;
        List<PratoCarrinho> pratoCarrinho = PratoCarrinho.findCarrinho(client, cliente).await().indefinitely();
        //Utilizar mapstruts
        List<PratoPedidoDTO> pratos = pratoCarrinho.stream().map(pc -> from(pc)).collect(Collectors.toList());
        pedido.pratos = pratos;

        RestauranteDTO restaurante = new RestauranteDTO();
        restaurante.nome = "nome restaurante";
        pedido.restaurante = restaurante;
        emitterPedido.send(pedido);
        return PratoCarrinho.delete(client, cliente);
    }

    private PratoPedidoDTO from(PratoCarrinho pratoCarrinho) {
        PratoDTO dto = Prato.findById(client, pratoCarrinho.prato).await().indefinitely();
        return new PratoPedidoDTO(dto.nome,dto.descricao,dto.preco);
    }
}
