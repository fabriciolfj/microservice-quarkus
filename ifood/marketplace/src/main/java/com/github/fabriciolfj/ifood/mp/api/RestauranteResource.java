package com.github.fabriciolfj.ifood.mp.api;

import com.github.fabriciolfj.ifood.mp.api.dto.PratoDTO;
import com.github.fabriciolfj.ifood.mp.api.dto.RestauranteDTO;
import com.github.fabriciolfj.ifood.mp.domain.model.Prato;
import com.github.fabriciolfj.ifood.mp.domain.model.Restaurante;
import io.smallrye.mutiny.Multi;
import io.vertx.mutiny.pgclient.PgPool;

import javax.inject.Inject;
import javax.ws.rs.Consumes;
import javax.ws.rs.GET;
import javax.ws.rs.Path;
import javax.ws.rs.PathParam;
import javax.ws.rs.Produces;
import javax.ws.rs.core.MediaType;

@Path("/restaurantes")
@Produces(MediaType.APPLICATION_JSON)
@Consumes(MediaType.APPLICATION_JSON)
public class RestauranteResource {

    @Inject
    PgPool pgPool;

    @GET
    public Multi<RestauranteDTO> findAll() {
        return Restaurante.findAll(pgPool);
    }

    @GET
    @Path("/{restauranteId}/pratos")
    public Multi<PratoDTO> findPratos(@PathParam("restauranteId") final Long id) {
        return Prato.findByRestaurante(id, pgPool);
    }
}
