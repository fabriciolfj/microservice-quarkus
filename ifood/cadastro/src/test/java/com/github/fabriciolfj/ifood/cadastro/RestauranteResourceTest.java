package com.github.fabriciolfj.ifood.cadastro;

import com.github.database.rider.cdi.api.DBRider;
import com.github.database.rider.core.api.configuration.DBUnit;
import com.github.database.rider.core.api.configuration.Orthography;
import com.github.database.rider.core.api.dataset.DataSet;
import com.github.fabriciolfj.ifood.cadastro.config.CadastroTestLifecycleManager;
import io.quarkus.test.common.QuarkusTestResource;
import io.quarkus.test.junit.QuarkusTest;
import org.approvaltests.Approvals;
import org.junit.jupiter.api.Test;

import static io.restassured.RestAssured.given;

@DBUnit(caseInsensitiveStrategy = Orthography.LOWERCASE)
@DBRider
@QuarkusTest
@QuarkusTestResource(CadastroTestLifecycleManager.class)
public class RestauranteResourceTest {

    @Test
    @DataSet("restaurantes-cenario-1.yml")
    public void testBuscarRestaurantes() {
        String resultado = given()
                .when().get("/restaurantes")
                .then()
                .statusCode(200)
                .extract().asString();

        Approvals.verifyJson(resultado);
    }
}
