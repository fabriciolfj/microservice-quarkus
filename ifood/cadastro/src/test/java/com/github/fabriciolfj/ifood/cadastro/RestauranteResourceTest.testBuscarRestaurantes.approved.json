[
  {
    "nomeFantasia": "Novo Restaurante",
    "proprietario": "ID do keyclock"
  },
  {
    "nomeFantasia": "Novo Restaurante 2",
    "proprietario": "ID do keyclock 2"
  }
]